product_multi_barcode/views/product_template_view.xml

```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <record id="product_template_form_view_multi_barcode" model="ir.ui.view">
+        <field name="name">product.template.form.multi.barcode</field>
+        <field name="model">product.template</field>
+        <field name="priority">16</field>
+        <field name="inherit_id" ref="product.product_template_only_form_view" />
+        <field name="arch" type="xml">
+            <field name="barcode" position="attributes">
+                <attribute name="readonly">True</attribute>
+            </field>
+            <field name="barcode" position="after">
+                <field
+                    name="barcode_ids"
+                    attrs="{'invisible': [('product_variant_count', '>', 1)]}"
+                >
+                    <tree string="Barcodes" editable="bottom">
+                        <field name="sequence" widget="handle" />
+                        <field name="name" />
+                    </tree>
+                </field>
```

Comment:
(continuing from previous comment)

Issues I see with this:
If a template has multiple variants, you wouldn't know to which variant the barcode belongs to.
If the template has only 1 variant, having a product_id here is useless, so..

I would try to find a way to:
- Display barcode's `product_id` if `product_variant_count > 1`
- Otherwise, set a default value for barcode's `product_id` to be the template's only variant and hide the field

So, maybe something like this (haven't tested it):

```suggestion
                <field
                    name="barcode_ids"
                    attrs="{'invisible': [('product_variant_count', '&gt;', 1)]}"
                >
                    <tree string="Barcodes" editable="bottom">
                        <field name="sequence" widget="handle" />
                        <field name="product_id"/>
                        <field name="name" />
                    </tree>
                </field>
                <field
                    name="barcode_ids"
                    attrs="{'invisible': [('product_variant_count', '&lt;=', 1)]}"
                    context="{'default_product_id': product_variant_id}"
                >
                    <tree string="Barcodes" editable="bottom">
                        <field name="sequence" widget="handle" />
                        <field name="product_id" invisible="1"/>
                        <field name="name" />
                    </tree>
                </field>
```



If it doesn't work maybe see how odoo's doing it for seller_ids:

https://github.com/odoo/odoo/blob/ee38f8e393e937c5694e6b75cd05b8a35d4f1624/addons/purchase/views/product_views.xml#L14-L17

Reply:
mmm, I get the idea but:
1. I'm not convinced yet that it's better on terms of usability
2. I'm not going to be able to play with the new approach. The change we did  till now was only to improve usability (with some hacks) but keeping the module behavior (if I'm not wrong).

If anyone interested on continuing the work, go ahead!

Thanks!